tests/serialize/json_test.py

```diff
@@ -0,0 +1,254 @@
+        'row_indices': ['1'],
+    },
+    {
+        'label': 'int(str:0)',
+        'kind': 'data_table_cell',
+        'value_type': 'builtins.int',
+        'row_types': ['builtins.str'],
+        'row_indices': ['0'],
+    },
+    {
+        'label': 'int(str:1)',
+        'kind': 'data_table_cell',
+        'value_type': 'builtins.int',
+        'row_types': ['builtins.str'],
+        'row_indices': ['1'],
+    },
+    {
+        'label': 'provide_series[str, float]',
+        'kind': 'p_series',
+        'value_type': 'builtins.float',
+        'row_dim': 'builtins.str',
+        'labels': ['0', '1'],
+    },
+    {
+        'label': 'table_cell(int(str:0))',
+        'kind': 'p_table_cell',
+    },
+    {
+        'label': 'table_cell(int(str:1))',
+        'kind': 'p_table_cell',
+    },
+]
+expected_serialized_edges_param_table = [
+    {'source': 'as_float', 'target': 'float(str:0)'},
+    {'source': 'as_float', 'target': 'float(str:1)'},
+    {'source': 'float(str:0)', 'target': 'provide_series[str, float]'},
+    {'source': 'float(str:1)', 'target': 'provide_series[str, float]'},
+    {'source': 'int(str:0)', 'target': 'as_float'},
+    {'source': 'int(str:1)', 'target': 'as_float'},
+    {'source': 'provide_series[str, float]', 'target': 'Series[str, float]'},
+    {'source': 'table_cell(int(str:0))', 'target': 'int(str:0)'},
+    {'source': 'table_cell(int(str:1))', 'target': 'int(str:1)'},
+]
+
+
+def test_serialize_param_table() -> None:
+    pl = sl.Pipeline([as_float])
+    pl.set_param_table(sl.ParamTable(str, {int: [3, -5]}))
+    graph = pl.build(sl.Series[str, float], handler=sl.HandleAsBuildTimeException())
+    tg = TaskGraph(graph=graph, keys=sl.Series[str, float])
+    res = tg.serialize()
+    check_serialized_graph(
+        res,
+        expected_serialized_nodes_param_table,
+        expected_serialized_edges_param_table,
+    )
+
+
+def test_serialize_ids_are_unique() -> None:
+    pl = sl.Pipeline([make_int_b, zeros, to_string], params={Int[A]: 3})
+    graph = pl.build(str, handler=sl.HandleAsBuildTimeException())
+    tg = TaskGraph(graph=graph, keys=str)
+    res = tg.serialize()
+    node_ids = [node['id'] for node in res['nodes']]
+    assert len(node_ids) == len(set(node_ids))
+
+
+def test_serialize_validate_schema() -> None:
+    pl = sl.Pipeline([make_int_b, zeros, to_string], params={Int[A]: 3})
+    graph = pl.build(str, handler=sl.HandleAsBuildTimeException())
+    tg = TaskGraph(graph=graph, keys=str)
+    res = tg.serialize()
+    schema = json_schema()
+    jsonschema.validate(res, schema)
+
+
+def test_serialize_validate_schema_param_table() -> None:
+    pl = sl.Pipeline([as_float])
+    pl.set_param_table(sl.ParamTable(str, {int: [3, -5]}))
+    graph = pl.build(sl.Series[str, float], handler=sl.HandleAsBuildTimeException())
+    tg = TaskGraph(graph=graph, keys=str)
+    res = tg.serialize()
+    schema = json_schema()
+    jsonschema.validate(res, schema)
```

Comment:
Along the lines of my first comment, I think it would be good if we could have something a bit like a round-trip test where we save the graph, then load it again and compute something with it and check that it's the same as what the original graph gives?

But this implies we have something that can load saved graphs...